boards: arm: Add Renesas Spider board

aaillet opened the pull request:
This pull request is introducing Generation 4 of Renesas R-Car SoCs with the S4 SoC as well as the Spider board into Zephyr.

Gen4 SoCs vs Gen3 (Zephyr POV) :
- Cortex-R52 is now used as Real Time core
- We are now able to use embedded ARM timer
- Some IPs are the same and will use existing drivers, sometime with few improvements (UART, GPIO...)
- Introduction of new IPs

S4 SoC that is equipping multiple Gen4 boards have some Cortex-A55s and a Cortex-R52 core.

Spider board is the reference board for S4 SoC, developpment oriented, this board is providing a lot of available IOs.

Operation of basic samples such as hello_world and blinky have been validated.

aaillet committed:
drivers: gpio: rcar: Add R-Car Gen4 support

Renesas Gen4 SoCs GPIO IPs are using one more
register comparing to Gen3 SoCs.

The new "INEN" register is used to enable general input.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

Commit:
drivers: pinctrl: Add R-Car Gen4 support

Renesas R-Car Gen4 is different from Gen3 regarding pinmux.

While Gen3 had only one base address to manage all pins,
Gen4 has one set of pinmux registers per GPIO banks.

We could expose one pinmux register per GPIO controllers,
but that would break potential compatibility with Linux
Device tree.

Instead create a reg_base array to parse all reg base from
device tree and identify proper base address based on the pin
definition.

This imply to add a pfc_base parameter to most of the pfc_rcar
function.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>
Signed-off-by: Pierre Marzin <pierre.marzin@iot.bzh>
Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
dts: arm: Add Renesas r8a779f0 support

r8a779f0 is also know as S4, this SoC is part of the Gen4 SoC series,
has 8 Cortex-A55 and a dual core lockstep Cortex-R52 processor.

SCIF0 is dedicated to Zephyr and SCIF3 to Linux.

**Control Domains**
IMPORTANT: This SoC is divided into two "domains":
- Application domain that contain some peripherals as well as A55 & R52 cores.
- Control domain that contain a G4MH/RH850 MCU and other peripherals

In order to access control domain peripherals such as gpio4-7 and CAN-FD
from application domain, the G4MH MCU has to release a protection
mechanism from control domain buses.

"Protected" controllers will be flagged in gen4 device trees,
warning users that they need to flash a custom G4MH firmware
to unlock access to these controllers.

**Clock controller**
This SoC clock controller is offering "domains"
for each world (Zephyr/Linux).

These domains are several "entry points" to the clock controller
which are arbitrated to avoid a world from turning off a clock needed
by another one.

We decided to use the same domain as Linux because the
security mechanism as to be implemented before accessing
another domain.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
boards: arm: Add Renesas Spider support

Renesas Spider board use a S4 SoC (r8a779f0).

Add basic support for UART, GPIO and clock control.

We are using SCIF0 as SCIF3 is used by Linux.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

pmarzin commented on dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:
```diff
@@ -0,0 +1,62 @@
+			compatible = "renesas,rcar-pfc";
+			reg = <0xe6050000 0x16c>, <0xe6050800 0x16c>,
+				  <0xe6051000 0x16c>, <0xe6051800 0x16c>,
+				  <0xffd90000 0x16c>, <0xffd90800 0x16c>,
```
il manque un patch ici, les adresses sont censé être 0xdfd90000 ... pas 0xf)

aaillet replied:
fixed !

pmarzin commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -10,13 +10,23 @@
+static const uint32_t reg_base[] = {
+	LISTIFY(DT_NUM_REGS(DT_DRV_INST(0)), PFC_REG_ADDRESS, (,), DT_DRV_INST(0))
+};
+
```
Is it a good idea (or mandatory ?) to add a #else with #error "Not supported" ?

aaillet replied:
I think that as we have full control on the code tree , there is no way that this driver can be called by another "SOC SERIE" board.
But even if not used, a protection can be added !

aaillet replied:
added a protection

pmarzin commented on dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:
```diff
@@ -0,0 +1,43 @@
+			status = "disabled";
+		};
+
+		/* Cant manage to use GPIOs from PFSS domain buses yet */
```
should we change the msg here has we know now how to use them ?

aaillet replied:
fixed

pmarzin commented on dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi:
```diff
@@ -0,0 +1,78 @@
+
+		sram0: memory@40040000 {
+			compatible = "mmio-sram";
+			reg = <0x40040000 0x1fc0000>;
```
not sure about the 0x1fc0000; need to check that size.

pmarzin replied:
<html><body>

H'00_EB200000 | ～ | H'00_EB2FFFFF | RT-SRAM0 | RAM
-- | -- | -- | -- | --

</body></html>

So size should be 0x100000 (=1MB) :) To set the same size of memory as the one really used with RT-SRAM.

aaillet replied:
fixed

pmarzin commented on boards/arm/rcar_spider/Kconfig.board:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2022 IoT.bzh
```
2023

pmarzin commented on boards/arm/rcar_spider/Kconfig.defconfig:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on boards/arm/rcar_spider/rcar_spider_cr52-pinctrl.dtsi:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2022 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on boards/arm/rcar_spider/rcar_spider_cr52.dts:
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2022 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on soc/arm/renesas_rcar/gen4/Kconfig.defconfig.r8a779f0:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2021 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on soc/arm/renesas_rcar/gen4/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,13 @@
+# Renesas R-Car Gen4 SoC line
+
+# Copyright (c) 2022 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on soc/arm/renesas_rcar/gen4/Kconfig.series:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2022 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on soc/arm/renesas_rcar/gen4/Kconfig.soc:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2022 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on soc/arm/renesas_rcar/gen4/linker.ld:
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2021 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on soc/arm/renesas_rcar/gen4/soc.c:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2022 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on soc/arm/renesas_rcar/gen4/soc.h:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2022 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on drivers/clock_control/clock_control_renesas_cpg_mssr.h:
```diff
@@ -41,6 +41,37 @@ static const uint16_t srcr[] = {
+	0x2C60, 0x2C64, 0x2C68, 0x2C6C,
+};
+
+/* SCIF clock */
```
Use same comment as GEN3 [ligne 43](https://github.com/iotbzh/zephyr/pull/43/commits/294e168bc43bfca8f303eef515062ca914060bcf#diff-a3879d78cc1f7eba632c4da477dc497eeca5ba01899f769c8c7413ec4bebcbe0R43)
```c
/* Peripherals Clocks */
#define S0D12_PER_CLK_RATE        66600000        /* SCIF */
```

aaillet replied:
fixed

pmarzin commented on dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:
```diff
@@ -1,14 +1,25 @@
 / {
 	soc {
+		/* Clock controller
+		 * Using domain 0 as linux
```
Same as previous commit

pmarzin commented on dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:
```diff
@@ -1,14 +1,25 @@
+		 */
+		cpg: clock-controller@e6150000 {
+			compatible = "renesas,r8a779f0-cpg-mssr";
+			reg = <0xe6150000 0x4000>;
```
Same as previous commit

aaillet replied:
explained in commit msg

pmarzin commented on dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi:
```diff
@@ -0,0 +1,78 @@
+		};
+
+		/* Clock controller
+		 * Using domain 0 as linux
```
What is explained in this comment ? Why talking about domain into clock part ?

aaillet replied:
explained in commit msg
not same domain as control/app domain, internal clock controller domains

pmarzin commented on dts/arm/renesas/rcar/gen4/rcar_gen4_cr52.dtsi:
```diff
@@ -0,0 +1,78 @@
+		 */
+		cpg: clock-controller@e6150000 {
+			compatible = "renesas,rcar-cpg-mssr";
+			reg = <0xe6150000 0x4000>;
```
`0x4000` is same as linux ? Looks to end at `0x3c00` in docs (without clock monitoring)

aaillet replied:
Same as linux : https://github.com/renesas-rcar/linux-bsp/blob/v5.10.147/rcar-5.2.0.rc6/arch/arm64/boot/dts/renesas/r8a779f0.dtsi#L368

pmarzin commented on dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:
```diff
@@ -10,6 +10,14 @@
+			compatible = "renesas,rcar-pfc";
+			reg = <0xe6050000 0x16c>, <0xe6050800 0x16c>,
+				  <0xe6051000 0x16c>, <0xe6051800 0x16c>,
+				  <0xdfd90000 0x16c>, <0xdfd90800 0x16c>,
```
Add a comment for address conversion over control domain ?

aaillet replied:
epxlained that last four registers were located in control domain of the soc

pmarzin commented on soc/arm/renesas_rcar/gen4/pfc_r8a779f0.c:
```diff
@@ -0,0 +1,604 @@
+/*
+ * Copyright (c) 2022-2023 IoT.bzh
```
2023

aaillet replied:
fixed

pmarzin commented on boards/arm/rcar_spider/rcar_spider_cr52.dts:
```diff
@@ -0,0 +1,64 @@
+&scif3 {
+	pinctrl-0 = <&scif3_data_tx_default &scif3_data_rx_default>;
+	pinctrl-names = "default";
+	status = "okay";
```
Should be `disabled` to let Linux alone ?

aaillet replied:
why not, as it it well commented

aaillet replied:
fixed
removed scif3 from spider board dts
kept it as soc level dtsi and in pinctrl board file
correctly commented soc (r8a779f0) dtsi to show which scif is used by which world

pmarzin commented on boards/arm/rcar_spider/rcar_spider_cr52_defconfig:
```diff
@@ -0,0 +1,15 @@
+CONFIG_SOC_SERIES_RCAR_GEN4=y
+CONFIG_BOARD_RCAR_SPIDER_CR52=y
+CONFIG_CLOCK_CONTROL=y
+#FIXME: What is the correct counter frequency ?
```
Should be fixed before MERGE

aaillet replied:
fixed

pmarzin commented on boards/arm/rcar_spider/doc/img/rcar_s4_block_diagram.jpg:
For me to know, is Zephyr github adding this type of file as LFS ?

aaillet replied:
As file, see some work to reduce image sizes some time ago here : https://github.com/zephyrproject-rtos/zephyr/pull/49024

pmarzin replied:
ok

pmarzin reviewed (commented):
Nice job :)

aaillet committed:
doc: Add Renesas Spider board documentation

Adding Spider board documentation based on Renesas official documentation
and following Zephyr guideline.

The documentation is describing the board and its current
Zephyr support.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
drivers: clock: rcar: Add r8a779f0 support

r8a779f0 has its own clock tree.
Gen4 SoCs common registers addresses have been added.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>
Signed-off-by: Pierre Marzin <pierre.marzin@iot.bzh>

aaillet committed:
drivers: pinctrl: rcar: Add r8a779f0 support

Enable PFC controller for r8a779f0 SoC.
Declare pin list for r8a779f0 SoC.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
soc: arm: reneas: Add r8a779f0 support

r8a779f0 SoC is part of the Renesas R-Car Gen4 SoC series.

This SoC has a dual core lockstep Cortex-R52 CPU.

L1 cache and branch prediction are enabled for Renesas RCar Gen4 SoCs.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet commented:
PR ready for upstream repository !
